File: gfg/march2024/25_print_n_bit_binary_numbers_with_1s_greater_than_0s.cpp
// INTUITION
// Use backtracking, fill ones and then zeros according to condition

// CODE 

class Solution {
    public:	
    void solve(int i, int countOne, int countZero, string num, int &n, vector<string> &res) {
        if(i >= n) {
            res.push_back(num);
            return;
        }
        
        num += '1';
        solve(i + 1, countOne + 1, countZero, num, n, res);
        num.pop_back();
        
        if(countOne > countZero) {
            num += '0';
            solve(i + 1, countOne, countZero + 1, num, n, res);
        }
    }
    
	vector<string> NBitBinary(int n) {
	    vector<string> res;
	    string num = "";
	    solve(0, 0, 0, num, n, res);
	    return res;
	}
};

File: gfg/march2024/python/25_print_n_bit_binary_numbers_with_1s_greater_than_0s.py
# INTUITION
# Use backtracking, fill ones and then zeros according to condition

# CODE 

class Solution:
    def solve(self, i, countOne, countZero, num, n, res) -> None:
        if i >= n:
            res.append(num)
            return

        num += '1'
        self.solve(i + 1, countOne + 1, countZero, num, n, res)
        num = num[:-1]

        if countOne > countZero:
            num += '0'
            self.solve(i + 1, countOne, countZero + 1, num, n, res)
                               
    def NBitBinary(self, n):
        res = []
        num = ""
        self.solve(0, 0, 0, num, n, res)
        return res